Packages/com.unity.inputsystem/InputSystem/InputAnalytics.cs

```diff
@@ -157,6 +224,384 @@ public struct ShutdownEventData
+            /// a call to <see cref="End()"/> this method has no effect.
+            /// </remarks>
+            public void Begin()
+            {
+                if (hasSession)
+                    return; // Session already started.
+                
+                m_SessionStart = currentTime;
+            }
+
+            /// <summary>
+            /// Ends the current session.
+            /// </summary>
+            /// <remarks>
+            /// If the session has not previously been started via a call to <see cref="Begin()"/> calling this
+            /// method has no effect.
+            /// </remarks>
+            public void End()
+            {
+                if (!hasSession)
+                    return; // No pending session
+                
+                // Make sure we register focus out if failed to capture or not invoked
+                if (hasFocus)
+                    RegisterEditorFocusOut();
+                
+                // Compute and record total session duration
+                var duration = currentTime - m_SessionStart;
+                m_Data.sessionDurationSeconds += (float)duration;
+                
+                // Send analytics event
+                runtime.SendAnalytic(this);
+                
+                // Reset to allow instance to be reused
+                Initialize(m_Data.kind);
+            }
+
+#if UNITY_EDITOR && UNITY_2023_2_OR_NEWER
+            public bool TryGatherData(out IAnalytic.IData data, out Exception error)
+#else
+            public bool TryGatherData(out IInputAnalyticData data, out Exception error)
+#endif
+            {
+                if (!isValid)
+                {
+                    data = null;
+                    error = new Exception("Unable to gather data without a valid session");
+                    return false;
+                }
+                
+                data = this.m_Data;
+                error = null;
+                return true; 
+            }
+            
+            private void Initialize(InputActionsEditorKind kind)
+            {
+                m_FocusStart = float.NaN;
+                m_SessionStart = float.NaN;
+                
+                m_Data = new InputActionsEditorSessionData(kind);
+            }
+
+            private bool ImplicitFocus()
+            {
+                if (!hasSession)
+                    return false;
+                if (!hasFocus)
+                    RegisterEditorFocusIn();
+                return true;
+            }
+
+            private readonly IInputRuntime m_InputRuntime;
+            private InputActionsEditorSessionData m_Data;
+            private double m_FocusStart;
+            private double m_SessionStart;
+            
+            private static IInputRuntime runtime => InputSystem.s_Manager.m_Runtime;
+            private bool hasFocus => !double.IsNaN(m_FocusStart);
```

Comment:
Is this IsNaN case checked for never having been set? I don't see where the nan would get set. Doe this work?

Reply:
Its set on line #583 within Initialize() above as well as in RegisterFocusOut() above and RegisterFocusIn() above, not sure what you are implying by working? Its used instead of making a nullable type to derimine if the session currently has focus and is set back to NaN in RegisterFocusOut after pending focus time has been accumulated.

Reply:
Ah yes I see it now - thank you I missed that initialization.